File: rust/src/problem_2.rs
use std::cmp;

fn compare(str1: String, str2: String) -> String
{
    /*
    This is internal function.
    This funciton accepts two string and returns
    the longest common prefix
    */

    let n1 = str1.len();
    let n2 = str2.len();
    let min_len = cmp::min(n1, n2);
    let mut prefix = String::from("");
    for i in 0..min_len
    {
        if str1.chars().nth(i).unwrap()  != str2.chars().nth(i).unwrap()
        {
            break;
        }
        prefix.push(str1.chars().nth(i).unwrap()); //add current char into prefix
    }
    return prefix;
}


fn LongestCommonPrefix<'a>(arr: &'a Vec<String >) -> String
{
    /*
    This function accepts a vector of Strings and returns the
    longest common prefix between those strings.

    Example:
    given vector of strings ["apple", "app", "aple", "appl"], the longest common  
    prefix is "ap".

    parameters:
    arr : vector of strings

    Return -> string: the longest common prefix between the vector of strings
    */
    let n = arr.len();
    let mut common = arr[0].to_owned();
    for i in 1..n
    {
        common = compare(common, arr[i].to_owned());
        if common == "" {break;}  //no prefix is found
    }
    return common;
}

//=================================================
pub fn run2()
{
    /*
    run this function in the main.rs by using below code:

    mod problem_2; 
    problem_2::run2();
    */
    let   arr: Vec<_> = vec!["apple", "app", "aplp", "appleor"].into_iter().map(String::from).collect();
    let out =  LongestCommonPrefix(&arr);

    if out == "" { println!("There is no common prefix");}
    else {println! ("The longest common prefix is: {}", out) }


}

File: HW_1/problem2.cpp
#include <iostream>

using namespace std;

string compare(string str1, string str2)
{
    /*
    This is internal function.
    This funciton accepts two string and returns
    the longest common prefix
    */
    int n1 = str1.length();
    int n2 = str2.length();
    int min_len = min(n1, n2);
    string prefix = "";
    for (int i  = 0; i < min_len; ++i)
    {
        if (str1[i] != str2[i]) break;
        prefix = prefix+str1[i];
    }
    return prefix;
}

string LongestCommonPrefix(string arr[], int n)
{
    /*
    This function accepts an array of strings and returns the
    longest common prefix between those strings.

    Example:
    given array of strings ["apple", "app", "aple", "appl"], the longest common prefix is "ap".

    parameters:
    arr : array of strings
    n   : size of the array

    Return -> string: the longest common prefix between the array of strings
    */

    string common = arr[0];
    for (int i = 1; i < n; i++)
    {
        common = compare(common, arr[i]);
        if (common == "") break; // no common prefix
    }
    return common;
}

int main()
{
    string arr[] = {"apple", "app", "aplp", "appleor"};
    int n = sizeof(arr)/sizeof(arr[0]); //length of the array

    string out = LongestCommonPrefix(arr, n);
    if (out == "") cout<<"There is no common prefix";
    else cout<<"The longest common prefix is: "<<out;
    return 0;
}
